### PR TITLE
GH-41470: [C++] Reuse deduplication logic for direct registration

### PR DESCRIPTION
**Pull request by bkietz**
### Rationale for this change

As observed in https://github.com/apache/arrow/pull/41309 a crossbow job on mac is failing due to duplicate registration of a factory for the file:// scheme

### What changes are included in this PR?

Deduplication of registered filesystem factories is applied to direct registration as well as when merging registries.

### Are these changes tested?

No, we just need to verify that the problematic crossbow job is repaired.

### Are there any user-facing changes?

No

* GitHub Issue: #41470

### CODE REVIEW

**Comment by bkietz** [comment]
@github-actions crossbow submit -g homebrew

**Comment by bkietz** [comment]
@github-actions crossbow submit -g homebrew

**Comment by vibhatha** [comment]
@github-actions crossbow submit *java*

**Comment by lidavidm** [comment]
@github-actions crossbow submit java-jars

File: dev/tasks/java-jars/github.yml [hunk 1]
```diff
@@ -80,7 +80,7 @@ jobs:
       fail-fast: false
       matrix:
         platform:
-          - { runs_on: ["macos-latest"], arch: "x86_64"}
+          - { runs_on: ["macos-13"], arch: "x86_64"}
```
Comment (lidavidm) [inline dev/tasks/java-jars/github.yml:83]: macos-latest == macos-14 now so if we want amd64 we have to explicitly request macos-13

**Review by lidavidm — APPROVED** [review]

**Comment by bkietz** [comment]
The error doesn't seem related to filesystem anymore: https://github.com/ursacomputing/crossbow/actions/runs/8903205653/job/24451424677#step:6:16247

I'm not sure what it means though

**Comment by vibhatha** [comment]
@lidavidm this seems to be from the recent maven build update?

**Comment by lidavidm** [comment]
We can merge this regardless and @vibhatha can file a new ticket.

**Comment by raulcd** [comment]
I will have to manually include: https://github.com/apache/arrow/pull/41466/commits/2fd3fcc6e9c2bbd4b105f14306f0ea8bd7d59cec into 16.1.0 in order to have a passing java-jars job as we don't want to add the other commits of this PR (the filesystems split should go into 17.0.0 instead of 16.1.0)
I plan to add the above as a manual commit along with the fix for: https://github.com/apache/arrow/pull/41578

**Pull request merged**